### ISSUE TITLE
[Improvement] Permissions contextualization

### ISSUE THREAD

**Issue by Kern046** [OWASP/rbac#60]
Hello people !

This issue is a kind of proposal about the permissions.

Currently, we are using two hierarchical trees : one for the roles, the other for the permissions.

It allows applications to have a better manipulation of the users rights, but I think it is not enough.

For example : I have the permission "read_article". If I want to set a group of articles private, I have to operate other controls to check if I can read the article. It is sad that RBAC alone cannot check it itself.

What I'm proposing : On another project, we put, above RBAC, some models and services to set the notion of context. Imagine that Article is an object. When we check the "read_article" permission, a new table allows us to know on which object the permission is granted (Project\Model\Article for example), and the IDs for which the permission is available.

I think it would be better if RBAC let the possibility to add such level of control on the permissions. That's why I create this issue, to be a base of discussion on eventual pull requests.

I'm not sure if my english was proper enough to make you understand the idea, don't hesitate to ask questions if this was not clear !

**Comment by abiusx** [comment]
Kern,
I think that is part of the application logic.
RBAC allows you to do that. There is no way to simply embed that into the library though, without too much bloating. 
Give me better example of y our use cases, and I’ll tell you how to implement it.
-A

**Comment by Kern046** [comment]
I'll try to explain myself better : 

I'm a member of a club. I want to create an event for this club in a REST-API way.

In my script, I will code something like that with the current system : 

``` php
$rbac = new PhpRbac\Rbac();
$clubId = $_POST['club_id'];

if($rbac->check('create_club_event', $myUserId))
{
    // Create the event for the club associated to $clubId
}
else
{
    // Access not granted
}
```

The thing is, if I put a random ID instead of my own club ID in the request, The event will be created anyway. As this behaviour is not wanted for the application, I have to add this statement : 

``` php
if($rbac->check('create_club_event', $myUserId) && $club->hasMember($myUserId))
```

I'd rather use this kind of function : 

``` php
if($rbac->checkContext('create_club_event', 'Project\Model\Club', $clubId, $myUserId))
```

It would be handled with a table with four columns (this is not an optimized way) : permission_id, context_type, context_id, user_id.

I will have a line like that : $createClubEventId, 'Project\Model\Club', $myClubId, $myUserId.

The function `$rbac->checkContext()` will return true if the user has the permission for the specific Club.

**Comment by abiusx** [comment]
You want to manage access on “data records”. This can not be done by a functional access control subsystem, because every data record has one corresponding access control rule, it needs to get updated along with the data, deleted, duplicated and so on.

You can do two things to resolve this:
1. If you have a limited number of clubs, add all of them in one subtree of permission tree. Then assign every user belonging to one of those, to the respective element. This will resolve everything.
2. If you have dynamic club management, you need to have a separate structure in your database for its access control. If its one to one, just add a field, but for more complex structures you need to do accordingly.

Basically there is no way you want to manage those with RBAC,  as you’d have to create a very complicated connection just to show that some data record is associated with some user. The purpose of RBAC is not that, but it still allows you to have hierarchy in your data record access control.
-A

**Comment by Kern046** [comment]
Ok, thank you for your time and for your answers :) !

**Comment by abiusx** [comment]
Let me know if you had any further questions!
